### PR TITLE
Get dependencies from original symbol or file lookup

### PR DESCRIPTION
**Pull request by artificial-aidan**
Background here: https://github.com/grpc/grpc/issues/32899

> Because most use cases will require also requesting the transitive dependencies of requested files, the queries will also return all transitive dependencies of the returned file.

The logic isn't my best work, so definitely open to input on that. The gist of it is to use the file descriptor protos that are returned in the dependency resolution, but the spec didn't specify an order, so I had to do extra logic around lookup. I may go look at how other languages to it, to see if they unofficially assume the order of dependencies will be correct.

This is just the async side, if the feedback is good I'm happy to port it to the sync side as well.

Another change I made was for the test files. I used an empty descriptor pool for the client creation, I found that the descriptors were already loaded, and the test's weren't actually exercising the lookup functionality. I believe this is because the test process loads the server, which loads the protos into its pool. By using an empty pool for the client, we force the exercising of the lookup mechanisms.

Additionally, an optional `skip_check_method_available` was added to the clients. This follows the above in trying to optimize the speed at which we can reflect and call a method.

### CODE REVIEW

**Comment by ViridianForge** [comment]
@artificial-aidan - just wanted to let you know that I'll give this a thorough review tomorrow night. Thanks for all this work!

File: src/grpc_requests/aio.py [hunk 1]
```diff
@@ -448,17 +455,21 @@ async def _get_service_names(self):
         services = tuple([s.name for s in resp.list_services_response.service])
         return services
 
-    async def get_file_descriptor_by_name(self, name):
+    async def get_file_descriptors_by_name(self, name):
```
Comment (ViridianForge) [inline src/grpc_requests/aio.py:458]: While I'll wager there's been limited engagement with the original method - I think for the sake of not breaking anyone who might be using it, we should keep the original method, but mark it deprecated with docs or a message pointing folks to use  your new method.

Same thing goes for the `get_file_descriptors_by_symbol`.

File: src/grpc_requests/aio.py [hunk 2]
```diff
@@ -468,7 +479,14 @@ def _is_descriptor_registered(self, filename):
         except KeyError:
             return False
 
-    async def _register_file_descriptor(self, file_descriptor):
+    # In practice it always seems like descriptors are returned in an order that makes sense for dependency
+    # registration, but i can't find a guarantee in the spec
+    # Because of this, go one by one and register, using the other returned descriptors as possible dependencies
```
Comment (ViridianForge) [inline src/grpc_requests/aio.py:484]:
```suggestion
    # Iterate over descriptors for registration, including returned descriptors as possible dependencies.
    # This is necessary as while in practice descriptors appear to be returned in an order that works for dependency
    # registration, this is not guaranteed in the reflection specification.
```

Just some suggested wordsmithing - I'm making the assumption the referred to spec is the GRPC reflection spec.

File: src/grpc_requests/aio.py [hunk 3]
```diff
@@ -218,6 +218,7 @@ def __init__(
         descriptor_pool=None,
         ssl=False,
         compression=None,
+        skip_check_method_available=False,
```
Comment (ViridianForge) [inline src/grpc_requests/aio.py:221]: Not opposed to this add, just not immediately clear on how it plays into this PR specifically.

File: src/grpc_requests/aio.py [hunk 4]
```diff
@@ -477,8 +495,20 @@ async def _register_file_descriptor(self, file_descriptor):
             )
             for dep_file_name in dependencies:
                 if not self._is_descriptor_registered(dep_file_name):
-                    dep_desc = await self.get_file_descriptor_by_name(dep_file_name)
-                    await self._register_file_descriptor(dep_desc)
+                    # First look for dependency in the passed in descriptors
+                    dep_desc = next(
+                        (x for x in file_descriptors if x.name == dep_file_name), None
+                    )
```
Comment (ViridianForge) [inline src/grpc_requests/aio.py:501]: If the goal here is simply to determine whether or not the dependency description is available in the list of existing file descriptors - do you strictly need to build the iterator, or can you simply use the set comprehension and call your follow up logic if the result is the empty set?

**Review by ViridianForge — COMMENTED** [review]
I have to admit - to some degree, the use case is a bit over my head - but I think I'm picking up enough to offer some starting feedback.

As always, thanks so much for the contribution!

File: src/grpc_requests/aio.py (hunk 3 above)
Comment (artificial-aidan) [reply src/grpc_requests/aio.py:221]: This PR came out of me doing some optimization for our usage. And this was another choke point I found. I can break it into another PR if need be.

File: src/grpc_requests/aio.py (hunk 1 above)
Comment (artificial-aidan) [reply src/grpc_requests/aio.py:458]: That makes sense. Will do

File: src/grpc_requests/aio.py (hunk 4 above)
Comment (artificial-aidan) [reply src/grpc_requests/aio.py:501]: I'll give that a shot

File: src/grpc_requests/aio.py (hunk 3 above)
Comment (ViridianForge) [reply src/grpc_requests/aio.py:221]: Nah, go ahead and leave it in, was just curious to the reason for the addition.

**Comment by artificial-aidan** [comment]
@ViridianForge Cleaned some stuff up, addressed comments, and added a sync implementation

File: src/grpc_requests/aio.py (hunk 4 above)
Comment (ViridianForge) [reply src/grpc_requests/aio.py:510]: Good thought - hadn't considered the short-circuit working out that way. Sounds like the way to go then!

**Review by ViridianForge — APPROVED** [review]
:ship: 
Looks good! If there's anything else you need to do to unblock your workflows, throw up an issue so we can track it, otherwise I'll start working on getting the next release out soon.

**Comment by artificial-aidan** [comment]
That's all for now! Thanks for reviewing it.

**Pull request merged**